gtsam/sfm/DsfTrackGenerator.h

```diff
@@ -0,0 +1,172 @@
+
+/**
+ * @file DsfTrackGenerator.h
+ * @date July 2022
+ * @author John Lambert
+ * @brief Identifies connected components in the keypoint matched graph.
+ */
+
+#pragma once
+#include <pybind11/stl.h>
+#include <Eigen/Core>
+
+#include <gtsam/base/DSFMap.h>
+#include <gtsam/geometry/Point2.h>
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <optional>
+#include <vector>
+
+namespace gtsam {
+
+typedef DSFMap<IndexPair> DSFMapIndexPair;
+typedef std::pair<size_t, size_t> ImagePair;
+typedef Eigen::MatrixX2i CorrespondenceIndices; // N x 2 array
+
+//struct Keypoints;
+using KeypointCoordinates = Eigen::MatrixX2d;
+
+
+struct Keypoints
+{
+  KeypointCoordinates coordinates;
+  // typedef'd for Eigen::VectorXd
+  boost::optional<gtsam::Vector> scales;
+  boost::optional<gtsam::Vector> responses;
+
+  Keypoints(const gtsam::KeypointCoordinates& coordinates): coordinates(coordinates) {}; // boost::none
+};
+
+using KeypointsList = std::vector<Keypoints>;
+using KeypointsVector = std::vector<Keypoints>; // TODO(johnwlambert): prefer KeypointsSet?
+using MatchIndicesMap = std::map<ImagePair, CorrespondenceIndices>;
+
+
+// @param camera index
+// @param 2d measurement
+// Implemented as named tuple, instead of std::pair (like SfmMeasurement in SfmTrack.h)
+struct NamedSfmMeasurement
+{
+  size_t i;
+  gtsam::Point2 uv;
+
+  NamedSfmMeasurement(size_t i, gtsam::Point2 uv) : i(i), uv(uv) {}
+};
+
+
+class SfmTrack2d
+{
+  private:
+    std::vector<NamedSfmMeasurement> measurements_;
+
+  public:
+    void addMeasurement(const NamedSfmMeasurement &m) {
+      measurements_.emplace_back(m);
+    }
+    std::vector<NamedSfmMeasurement> measurements() {return measurements_; }
+
+  // @brief Validates the track by checking that no two measurements are from the same camera.
+  //
+  // returns boolean result of the validation.
+  bool validate_unique_cameras()
+  {
+    std::vector<int> track_cam_idxs;
+    for (auto & measurement: measurements_)
+    {
+      track_cam_idxs.emplace_back(measurement.i);
+    }
+    auto i = std::adjacent_find(track_cam_idxs.begin(), track_cam_idxs.end());
+    bool all_cameras_unique = (i == track_cam_idxs.end());
+    return all_cameras_unique;
+  }
+};
+
+
+class DsfTrackGenerator
+{
+
+  public:
+    DsfTrackGenerator() {}
+
+    // Creates a list of tracks from 2d point correspondences.
+    //
+    // Creates a disjoint-set forest (DSF) and 2d tracks from pairwise matches. We create a singleton for union-find
+    // set elements from camera index of a detection and the index of that detection in that camera's keypoint list,
+    // i.e. (i,k).
+    std::vector<SfmTrack2d> generate_tracks_from_pairwise_matches(
+        const MatchIndicesMap& matches_dict,
+        const KeypointsList& keypoints_list)
+    {
+      std::vector<SfmTrack2d> track_2d_list;
+
+      std::cout << "[SfmTrack2d] Starting Union-Find..." << std::endl;
```

Comment:
the union find is an implementation detail I think, no need to log that.. 
Maybe this function does not need any std::cout statements. You could leave the one that prints the number of tracks if you'd like.